h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/matrix/DenseMatrixFactory.java

```diff
@@ -51,8 +51,40 @@ protected DenseDMatrixProvider(
             super(actualRows, response, weights, offsets);
             this.data = data;
         }
+        
+        @Override
+        public void print(int nrow) {
+            for (int i = 0; i < (nrow > 0 ? nrow : data.nrow); i++) {
+                System.out.print(i + ":");
+                for (int j = 0; j < data.ncol; j++) {
+                    System.out.print(data.get(i, j) + ", ");
+                }
+                System.out.print(response[i]);
+                System.out.println();
+            }
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            if (!super.equals(o)) return false;
+            DenseDMatrixProvider that = (DenseDMatrixProvider) o;
+            if (that.data.ncol != this.data.ncol || that.data.nrow != this.data.nrow) return false;
+            for (int i = 0; i < this.data.nrow; i++) {
+                for (int j = 0; j < this.data.ncol; j++) {
+                    if (this.data.get(i, j) != that.data.get(i, j)) {
+                        return false;
+                    }
+                }
+            }
+            return true;
+        }
 
-        public DenseDMatrixProvider() {}
+        @Override
+        public int hashCode() {
+            return super.hashCode() + Objects.hash(data);
```

Comment:
2 comments:
1) BigDenseMatrix doesn't implement hashCode function - is that an issue
2) Using Objects.hash, in this case, is inefficient, a new array to wrap the single element needs to be created, it then essentially just does something like `data == null ? 0 : 31 * data.hashCode()`